gk/main.c

```diff
@@ -806,6 +820,105 @@ gk_setup_rss(struct gk_config *gk_conf)
 	return ret;
 }
 
+/*
+ * Consult the RED algorithm to determine whether this packet should
+ * be marked or dropped.
+ *
+ * If @encapsulated is true, then @pkt is an IP-in-IP packet, and both
+ * the outer and inner IP headers ECN bits should be set. This function
+ * assumes that the encapsulation is either IPv4-in-IPv4 or IPv6-in-IPv6,
+ * as that's what encapsulate() does.
+ *
+ * Returns -1 if the packet is to be dropped or 0 if it should be enqueued.
+ */
+static int
+red_check(struct gk_config *gk_conf, struct gk_instance *instance,
+	struct rte_mbuf *pkt, struct ipacket *packet, int encapsulated,
+	uint16_t queue_len, uint64_t tx_bytes)
+{
+	int ret = rte_red_enqueue(&gk_conf->red_cfg,
+		&instance->red, queue_len, tx_bytes);
+
+	/* Max threshold exceeded or drop by mark probability. */
+	if (ret == 1 || (ret == 2 && instance->red.count == 0)) {
+		rte_pktmbuf_free(pkt);
+		return -1;
+	}
+
+	/* Average queue length is below minimum threshold. */
+	if (ret == 0)
+		goto enqueue;
+
+	/*
+	 * The queue length is above the minimum threshold but the
+	 * packet was not dropped. Mark the ECN bits if needed.
+	 */
+
+	RTE_VERIFY(ret == 2);
+
+	/*
+	 * Either a router between the source and this Gatekeeper
+	 * server already marked the ECN bits or the sender is not ECN
+	 * enabled. In either case, we copied the ECN bits to the outer
+	 * header already in encapsulate(), so no need to continue.
+	 */
+	if (packet->ecn == IPTOS_ECN_CE || packet->ecn == IPTOS_ECN_NOT_ECT)
+		goto enqueue;
+
+	if (packet->flow.proto == ETHER_TYPE_IPv4) {
+		/*
+		 * If @encapsulated is false, then @outer_ip4hdr is the
+		 * only available IPv4 header and @inner_ip4hdr is not used.
+		 */
+		struct ipv4_hdr *outer_ip4hdr, *inner_ip4hdr;
+		unsigned long sum;
+		uint8_t old_tos;
+
+		outer_ip4hdr = rte_pktmbuf_mtod_offset(pkt, struct ipv4_hdr *,
+			sizeof(struct ether_hdr));
+
+		/* Mark as congested by setting both ECN bits. */
+		outer_ip4hdr->type_of_service |= IPTOS_ECN_CE;
+
+		if (!encapsulated)
+			goto enqueue;
+
+		/*
+		 * In IPv4, need to recompute the header checksum.
+		 * Use the incremental checksum update algorithm
+		 * as shown in RFC 1141.
+		 */
+		inner_ip4hdr = &outer_ip4hdr[1];
```

Comment:
Yes, thanks -- we can just wait to copy it when we decapsulate it at Grantor. (And should be changing the IPv4 checksum there, too.)